File: go/runtime.go
package rvgen

import (
	"encoding/binary"
	"fmt"
	"rvgen/riscv"
)

func pcRelative(delta int64) ([2]uint32, error) {
	hi := (delta + 0x800) >> 12
	if hi < -0x80000 || hi > 0x7ffff {
		return [2]uint32{}, fmt.Errorf("runtime target exceeds AUIPC range")
	}
	return [2]uint32{riscv.Rv32i_auipc(5, int32(hi)), riscv.Rv32i_addi(5, 5, int32(delta-(hi<<12)))}, nil
}
func appendWords(dst []byte, words ...uint32) []byte {
	for _, w := range words {
		dst = binary.LittleEndian.AppendUint32(dst, w)
	}
	return dst
}

// Executable wraps a workload in a machine-mode trap handler and HTIF exit
// runtime. Normal completion writes 1 to tohost; traps write 3 (failure).
func Executable(body []byte, is64 bool, start uint64) ([]byte, error) {
	if start&3 != 0 || len(body)&1 != 0 {
		return nil, fmt.Errorf("entry must be four-byte aligned and body must contain whole instructions")
	}
	text := make([]byte, 8, len(body)+64)
	text = appendWords(text, riscv.Zicsr_csrrw(0, 5, riscv.CSR_MTVEC))
	text = append(text, body...)
	if len(text)&3 != 0 {
		text = append(text, 1, 0)
	}
	text = appendWords(text, riscv.Rv32i_addi(31, 0, 1))
	writerOffset := len(text)
	text = append(text, make([]byte, 8)...)
	store := riscv.Rv64i_sd(5, 31, 0)
	if !is64 {
		text = appendWords(text, riscv.Rv32i_sw(5, 0, 4))
		store = riscv.Rv32i_sw(5, 31, 0)
	}
	text = appendWords(text, riscv.Rv32i_fence(0x33), store, riscv.Rv32i_fence(0x33), riscv.Rv32i_jal(0, 0))
	trapOffset := len(text)
	text = appendWords(text, riscv.Rv32i_addi(31, 0, 3), riscv.Rv32i_jal(0, int32(writerOffset-trapOffset-4)))
	textEnd, err := add64(start, uint64(len(text)))
	if err != nil {
		return nil, err
	}
	tohostAddr, err := alignUp(textEnd, 64)
	if err != nil {
		return nil, err
	}
	fromhostAddr, err := add64(tohostAddr, 64)
	if err != nil {
		return nil, err
	}
	entryPair, err := pcRelative(int64(trapOffset))
	if err != nil {
		return nil, err
	}
	writerPair, err := pcRelative(int64(tohostAddr-start) - int64(writerOffset))
	if err != nil {
		return nil, err
	}
	for i, w := range entryPair {
		binary.LittleEndian.PutUint32(text[i*4:], w)
	}
	for i, w := range writerPair {
		binary.LittleEndian.PutUint32(text[writerOffset+i*4:], w)
	}
	code := NewELFSection(".text", text)
	code.Addr = start
	tohost := ELFSection{Name: ".tohost", Bytes: make([]byte, 8), Addr: tohostAddr, Flags: SHFAlloc | SHFWrite, Align: 64, Type: 1}
	fromhost := tohost
	fromhost.Name = ".fromhost"
	fromhost.Addr = fromhostAddr
	symbols := []ELFSymbol{
		{"_start", 0, 0, uint64(len(text)), 2}, {"rvgen_trap", 0, uint64(trapOffset), 8, 2},
		{"tohost", 1, 0, 8, 1}, {"fromhost", 2, 0, 8, 1},
	}
	return BuildELFWithSymbols([]ELFSection{code, tohost, fromhost}, symbols, is64, start)
}

File: src/runtime.rs
//! Bare-metal entry, trap handling, and Spike HTIF termination.
//!
//! HTIF discovers the `tohost` and `fromhost` ELF symbols and treats an odd
//! value written to `tohost` as `(exit_status << 1) | 1`. Normal completion
//! writes 1 to `tohost`; traps write 3 (failure).

use std::path::Path;

use anyhow::{Context, Result, bail};

use crate::riscv::{Csr, Instruction, XReg};
use crate::utils::{ElfSection, ElfSymbol, SHF_ALLOC, SHF_WRITE, build_elf_with_symbols};

pub fn pc_relative(delta: i64) -> Result<[u32; 2]> {
    let hi = (delta + 0x800) >> 12;
    if !(-0x80000..=0x7ffff).contains(&hi) {
        bail!("runtime target exceeds AUIPC range");
    }
    let lo = (delta - (hi << 12)) as i32;
    let auipc = Instruction::Auipc {
        rd: XReg::X5,
        imm: hi as i32,
    }
    .encode()?
    .bits();
    let addi = Instruction::Addi {
        rd: XReg::X5,
        rs1: XReg::X5,
        imm: lo,
    }
    .encode()?
    .bits();
    Ok([auipc, addi])
}

fn append_words(dst: &mut Vec<u8>, words: &[u32]) {
    for &w in words {
        dst.extend_from_slice(&w.to_le_bytes());
    }
}

/// Wraps a workload body in a machine-mode trap handler and HTIF exit runtime,
/// producing a complete ELF executable. Initializes integer registers and FPU
/// control state before entering the body; requires the F extension.
fn executable(body: &[u8], is_64bit: bool, start_addr: u64) -> Result<Vec<u8>> {
    if start_addr & 3 != 0 || body.len() & 1 != 0 {
        bail!("entry must be four-byte aligned and body must contain whole instructions");
    }

    // Reserve 8 bytes at the start for AUIPC/ADDI that point to the trap handler
    let mut text = vec![0u8; 8];
    let csrrw = Instruction::Csrrw {
        rd: XReg::ZERO,
        rs1: XReg::X5,
        csr: Csr::MTVEC,
    }
    .encode()?
    .bits();
    append_words(&mut text, &[csrrw]);

    // Enable the FPU before accessing fcsr. Set both mstatus.FS bits (Dirty)
    // so the final state does not depend on the boot code's previous FS value.
    // CSRRS preserves the other mstatus fields.
    for instr in [
        Instruction::Lui {
            rd: XReg::X5,
            imm: 0x6, // 0x6000 = 0b11 << 13 (mstatus.FS)
        },
        Instruction::Csrrs {
            rd: XReg::ZERO,
            rs1: XReg::X5,
            csr: Csr::MSTATUS,
        },
        // frm = RNE (round to nearest, ties to even), fflags = 0.
        Instruction::Csrrw {
            rd: XReg::ZERO,
            rs1: XReg::ZERO,
            csr: Csr::FCSR,
        },
    ] {
        instr.append_bytes(&mut text)?;
    }

    // x0 is hardwired to zero. Clear x1-x31 after all setup so scratch
    // registers (including x5) cannot leak boot/runtime state into the body.
    for index in 1..32 {
        Instruction::Addi {
            rd: XReg::new(index)?,
            rs1: XReg::ZERO,
            imm: 0,
        }
        .append_bytes(&mut text)?;
    }

    // Append workload body
    text.extend_from_slice(body);

    // Pad with c.nop if 2-byte aligned
    if text.len() & 3 != 0 {
        text.extend_from_slice(&[1, 0]); // c.nop
    }

    // Normal exit: x31 = 1 (HTIF success)
    let addi_pass = Instruction::Addi {
        rd: XReg::X31,
        rs1: XReg::ZERO,
        imm: 1,
    }
    .encode()?
    .bits();
    append_words(&mut text, &[addi_pass]);

    let writer_offset = text.len();
    // Reserve 8 bytes for AUIPC/ADDI that calculate tohost address into x5
    text.extend_from_slice(&[0u8; 8]);

    let store = if is_64bit {
        Instruction::Sd {
            rs1: XReg::X5,
            rs2: XReg::X31,
            imm: 0,
        }
        .encode()?
        .bits()
    } else {
        Instruction::Sw {
            rs1: XReg::X5,
            rs2: XReg::X31,
            imm: 0,
        }
        .encode()?
        .bits()
    };

    let fence = Instruction::Fence { imm: 0x33 }.encode()?.bits();
    let jal_spin = Instruction::Jal {
        rd: XReg::ZERO,
        imm: 0,
    }
    .encode()?
    .bits();

    if !is_64bit {
        let sw_upper = Instruction::Sw {
            rs1: XReg::X5,
            rs2: XReg::ZERO,
            imm: 4,
        }
        .encode()?
        .bits();
        append_words(&mut text, &[sw_upper]);
    }

    append_words(&mut text, &[fence, store, fence, jal_spin]);

    // Trap handler: x31 = 3 (HTIF failure)
    let trap_offset = text.len();
    let addi_fail = Instruction::Addi {
        rd: XReg::X31,
        rs1: XReg::ZERO,
        imm: 3,
    }
    .encode()?
    .bits();
    let jal_to_writer = Instruction::Jal {
        rd: XReg::ZERO,
        imm: (writer_offset as i64 - trap_offset as i64 - 4) as i32,
    }
    .encode()?
    .bits();
    append_words(&mut text, &[addi_fail, jal_to_writer]);

    let text_end = start_addr
        .checked_add(text.len() as u64)
        .ok_or_else(|| anyhow::anyhow!("runtime address overflow"))?;
    let tohost_addr = (text_end
        .checked_add(63)
        .ok_or_else(|| anyhow::anyhow!("runtime address overflow"))?)
        & !63;
    let fromhost_addr = tohost_addr
        .checked_add(64)
        .ok_or_else(|| anyhow::anyhow!("runtime address overflow"))?;

    let entry_pair = pc_relative(trap_offset as i64)?;
    let writer_pair = pc_relative((tohost_addr - start_addr) as i64 - writer_offset as i64)?;

    for (i, &w) in entry_pair.iter().enumerate() {
        text[i * 4..i * 4 + 4].copy_from_slice(&w.to_le_bytes());
    }
    for (i, &w) in writer_pair.iter().enumerate() {
        text[writer_offset + i * 4..writer_offset + i * 4 + 4].copy_from_slice(&w.to_le_bytes());
    }

    let text_len = text.len() as u64;
    let mut code_section = ElfSection::new(".text", text);
    code_section.addr = start_addr;

    let mut tohost_section = ElfSection::new(".tohost", vec![0u8; 8]);
    tohost_section.addr = tohost_addr;
    tohost_section.flags = SHF_ALLOC | SHF_WRITE;
    tohost_section.align = 64;

    let mut fromhost_section = tohost_section.clone();
    fromhost_section.name = ".fromhost".to_string();
    fromhost_section.addr = fromhost_addr;

    let symbols = [
        ElfSymbol {
            name: "_start".to_string(),
            section: 0,
            offset: 0,
            size: text_len,
            symbol_type: 2, // STT_FUNC
        },
        ElfSymbol {
            name: "rvgen_trap".to_string(),
            section: 0,
            offset: trap_offset as u64,
            size: 8,
            symbol_type: 2, // STT_FUNC
        },
        ElfSymbol {
            name: "_exit".to_string(),
            section: 0,
            offset: (trap_offset - 4) as u64,
            size: 0,
            symbol_type: 2, // STT_FUNC
        },
        ElfSymbol {
            name: "tohost".to_string(),
            section: 1,
            offset: 0,
            size: 8,
            symbol_type: 1, // STT_OBJECT
        },
        ElfSymbol {
            name: "fromhost".to_string(),
            section: 2,
            offset: 0,
            size: 8,
            symbol_type: 1, // STT_OBJECT
        },
    ];

    build_elf_with_symbols(
        &[code_section, tohost_section, fromhost_section],
        &symbols,
        is_64bit,
        start_addr,
    )
}

/// Convenience function to generate an executable ELF and write it directly to a file.
pub fn write_executable(
    body: &[u8],
    is_64bit: bool,
    start_addr: u64,
    path: impl AsRef<Path>,
) -> Result<()> {
    let elf_data = executable(body, is_64bit, start_addr)?;
    let p = path.as_ref();
    std::fs::write(p, elf_data).with_context(|| format!("failed to write ELF to {}", p.display()))
}
